include/nbl/builtin/hlsl/type_traits.hlsl

```diff
@@ -4,13 +4,425 @@
+#define NBL_REGISTER_TYPEID(T) template<> struct typeid_t<T> : integral_constant<uint32_t,__COUNTER__> {}
+
+#define NBL_REGISTER_OBJ_TYPE(T) NBL_REGISTER_TYPEID(T); \
+namespace impl \
+{ \
+encoder<typeid_t<T>::value> encode_typeid(T); \
+template<> struct decltype_t<sizeof(encoder<typeid_t<T>::value>)/4> \
+{ \
+  using type = T; \
+}; \
+} 
+
+#define typeid(expr) (sizeof(::nbl::hlsl::type_traits::impl::encode_typeid(expr))/4)
+#define decltype(expr) ::nbl::hlsl::type_traits::impl::decltype_t<typeid(expr)>::type
+
+// builtins
+NBL_REGISTER_OBJ_TYPE(bool)
+NBL_REGISTER_OBJ_TYPE(int16_t)
+NBL_REGISTER_OBJ_TYPE(int32_t)
+NBL_REGISTER_OBJ_TYPE(int64_t)
+NBL_REGISTER_OBJ_TYPE(uint16_t)
+NBL_REGISTER_OBJ_TYPE(uint32_t)
+NBL_REGISTER_OBJ_TYPE(uint64_t)
+NBL_REGISTER_OBJ_TYPE(float32_t)
+NBL_REGISTER_OBJ_TYPE(float64_t)
+
+#else // C++
+
+template<class T, T val>
+using integral_constant = std::integral_constant<T, val>;
+
+template<bool val>
+using bool_constant = std::bool_constant<val>;
+
+template <bool C, class T, class F>
+using conditional  = std::conditional<C, T, F>;
+
+template<class A, class B>
+using is_same = std::is_same<A, B>;
+
+template<class T>
+using is_void = std::is_void<T>;
+
+template<class T>
+using is_array = std::is_array<T>;
+
+template<class T>
+using is_scalar = std::is_scalar<T>;
+
+template<class T>
+struct is_signed : impl::base_type_forwarder<std::is_signed, T> {};
+
+template<class T>
+struct is_unsigned : impl::base_type_forwarder<std::is_unsigned, T> {};
+
+template<class T>
+struct is_integral : impl::base_type_forwarder<std::is_integral, T> {};
+
+template<class T>
+struct is_floating_point : impl::base_type_forwarder<std::is_floating_point, T> {};
+
+template<class T>
+using is_const = std::is_const<T>;
+
+template<class T>
+using is_volatile = std::is_volatile<T>;
+
+template<class T>
+using is_trivial = std::is_trivial<T>;
+
+template<class T>
+using is_trivially_copyable = std::is_trivially_copyable<T>;
+
+template<class T> 
+using is_empty = std::is_empty<T>;
+
+template<class T> 
+using is_polymorphic = std::is_polymorphic<T>;
+
+template<class T> 
+using is_abstract = std::is_abstract<T>;
+
+template<class T> 
+using is_final = std::is_final<T>;
+
+template<class T>
+using is_fundamental = std::is_fundamental<T>;
+
+template<class T>
+using is_compound = std::is_compound<T>;
+
+template<class T>
+using is_aggregate = std::is_aggregate<T>;
+
+template<typename T>
+struct typeid_t : std::integral_constant<uint64_t,typeid(T).hash_code()> {};
+
+#endif
+
+// Overlapping definitions
+
+template<class T>
+struct is_vector : bool_constant<false> {};
+
+template<class T>
+struct is_matrix : bool_constant<false> {};
+
+template<class T, uint32_t N>
+struct is_vector<vector<T, N> > : bool_constant<true> {};
+
+template<class T, uint32_t N, uint32_t M>
+struct is_matrix<matrix<T, N, M> > : bool_constant<true> {};
```

Comment:
where's `scalar_type` from @Przemog1's `type_traits` ?

Reply:
oh, i thought it was on master let me merge that real quick